### ISSUE TITLE
Nginx Server with QUICHE: - QLogs option

### ISSUE THREAD

**Issue by iamPerfPro** [cloudflare/quiche#869]
I 'm able to run Nginx with the QUICHE patch and server pages over both H2 and H3. Is there a way to enable qlogs switch so that Qlogs can generated by NGINX when serving webpages on H3 using QUICHE with Nginx as server. 

Any guidance would be great.

**Comment by LPardue** [comment]
It's possible yes. At the most basic level, on Linux you can just call `quiche_conn_set_qlog_fd()` in ngx_event_quic.c:ngx_quic_create_connection()`

You need to decide on what file descriptor you want to pass, and that's the more complex bit. But for a quick fix I'd recommend trying it out with a file named after the CID.

I've got code that handles things more elegantly but I've never had the time to polish up  enough get it landed in our patch.

**Comment by abnv418** [comment]
Thanks for your reply. 
I start the nginx like 
./nginx.c -c ngin.conf 

and I run the quiche server using the following command: -

cargo run --manifest-path=tools/apps/Cargo.toml --bin quiche-server -- \
--cert tools/apps/src/bin/cert.crt \
--key tools/apps/src/bin/cert.key \
--root /Desktop/root/ \
--index index.html

After setting the above where should I give the parameters for q log so that Nginx server gives qlog at a location. ?

Thanks

**Comment by LPardue** [comment]
Yeah this is where all the complexity comes from. My patch aligned with other logging configuration strategies. 

If you're patching something yourself in the short term, I'd just suggest doing the most simple thing.

**Comment by abnv418** [comment]
Hi thanks again for your reply. Can you please guide me more about when you say: -  "recommend trying it out with a file named after the CID."  What is CID here. I'm little confused with that.

**Comment by abnv418** [comment]
I tried to call the function like 
quiche_conn_set_qlog_fd(conn,100,description,title); 

by setting the variables as 

char 	*description= "description";
char         *title= "title";

but I get this error when I call it. Is there any lead as to why is this happening : - 
/usr/bin/ld: objs/src/event/ngx_event_quic.o: in function `ngx_quic_create_connection':
Desktop/newnginx/nginx-1.16.1/src/event/ngx_event_quic.c:191: undefined reference to `quiche_conn_set_qlog_fd'
/usr/bin/ld: ../quiche/target/release/libquiche.a(quiche.quiche.beisisz9-cgu.0.rcgu.o): in function `quiche::tls::Handshake::use_legacy_codepoint':

/Desktop/newnginx/quiche/src/tls.rs:343: undefined reference to `SSL_set_quic_use_legacy_codepoint'
/usr/bin/ld: /home/parallels/Desktop/newnginx/quiche/src/tls.rs:343: undefined reference to `SSL_set_quic_use_legacy_codepoint'
/usr/bin/ld: ../quiche/target/release/libquiche.a(quiche.quiche.beisisz9-cgu.11.rcgu.o): in function `quiche::tls::Handshake::use_legacy_codepoint':
/Desktop/newnginx/quiche/src/tls.rs:343: undefined reference to `SSL_set_quic_use_legacy_codepoint'
collect2: error: ld returned 1 exit status
make[1]: *** [objs/Makefile:260: objs/nginx] Error 1
make[1]: Leaving directory '/home/parallels/Desktop/newnginx/nginx-1.16.1'
make: *** [Makefile:8: build] Error 2

**Comment by LPardue** [comment]
What OS are you using? I think the fd option only works on Linux.

**Comment by abnv418** [comment]
I'm using Ubuntu 20.04.2 LTS.

**Comment by LPardue** [comment]
ah you might need to enable the `qlog` feature when building quiche otherwise its not included and the linking will fail.

**Comment by abnv418** [comment]
So the build is working with qlog feature enabled. Can you tell me the process to enable the normal logging in QUICHE. I can see via nginx logs the execution is hitting quiche_conn_set_qlog_fd() function. But nothing gets printed in the qlog file.  So I wanted to debug the QUICHE side. Can you guide how to enable the logging on QUICHE when used after patching with  nginx.

**Comment by Ryenum** [comment]
您好，我在配置 NGINX 的 quic 服务时遇到了一些问题。我的配置和官网一样，但是访问服务器时还是不能使用quic协议。h2 协议仍在使用。这是我的编译信息： root@learn：/usr/local/nginx/sbin# ./nginx -V
nginx version：
nginx/1.16.1 （quiche-83d9168a） 由 gcc 11.4.0 （Ubuntu 11.4.0-1ubuntu1~22.04）

构建 OpenSSL 1.1.1 构建（兼容;BoringSSL） （使用 BoringSSL 运行）
已启用
TLS SNI 支持 配置参数： --prefix=/usr/local/nginx --build=quiche-83d9168a --with-http_v3_module --with-http_v2_module --with-quiche=/quiche --with-http_ssl_module --with-openssl=/quiche/quiche/deps/boringssl --add-module=/connect

**Comment by Ryenum** [comment]
Hello, I encountered some problems when configuring the quic service of NGINX. My configuration is the same as that of the official website, but I still cannot use the quic protocol when accessing the server. The h2 protocol is still used. Here is my compilation information:
nginx version: nginx/1.16.1 (quiche-83d9168a)
built by gcc 11.4.0 (Ubuntu 11.4.0-1ubuntu1~22.04) 
built with OpenSSL 1.1.1 (compatible; BoringSSL) (running with BoringSSL)
TLS SNI support enabled
configure arguments: --prefix=/usr/local/nginx --build=quiche-83d9168a --with-http_v3_module --with-http_v2_module --with-quiche=/quiche --with-http_ssl_module --with-openssl=/quiche/quiche/deps/boringssl --add-module=/connect

**Comment by LPardue** [comment]
As of https://github.com/cloudflare/quiche/commit/e434e42173f73bb533f6a39cd853eac752731fd9 the nginx patch was removed from quiche. Hence we're closing this issue.

**Comment by Adamkaram** [comment]
why we remove it ?
